cpp/src/arrow/compute/kernels/sum-internal.h

```diff
@@ -0,0 +1,204 @@
+#include "arrow/type.h"
+#include "arrow/type_traits.h"
+#include "arrow/util/bit-util.h"
+
+namespace arrow {
+
+class Array;
+class DataType;
+
+namespace compute {
+
+// Find the largest compatible primitive type for a primitive type.
+template <typename I, typename Enable = void>
+struct FindAccumulatorType {};
+
+template <typename I>
+struct FindAccumulatorType<I, enable_if_signed_integer<I>> {
+  using Type = Int64Type;
+};
+
+template <typename I>
+struct FindAccumulatorType<I, enable_if_unsigned_integer<I>> {
+  using Type = UInt64Type;
+};
+
+template <typename I>
+struct FindAccumulatorType<I, enable_if_floating_point<I>> {
+  using Type = DoubleType;
+};
+
+template <typename ArrowType, typename StateType>
+class SumAggregateFunction final : public AggregateFunctionStaticState<StateType> {
+  using CType = typename TypeTraits<ArrowType>::CType;
+  using ArrayType = typename TypeTraits<ArrowType>::ArrayType;
+
+  static constexpr int64_t kTinyThreshold = 32;
+  static_assert(kTinyThreshold > 18,
+                "ConsumeSparse requires at least 18 elements to fit 3 bytes");
+
+ public:
+  Status Consume(const Array& input, StateType* state) const override {
+    const ArrayType& array = static_cast<const ArrayType&>(input);
+
+    if (input.null_count() == 0) {
+      *state = ConsumeDense(array);
+    } else if (input.length() <= kTinyThreshold) {
+      // In order to simplify ConsumeSparse implementation (requires at least 3
+      // bytes of bitmap data), small arrays are handled differently.
+      *state = ConsumeTiny(array);
+    } else {
+      *state = ConsumeSparse(array);
+    }
+
+    return Status::OK();
+  }
+
+  Status Merge(const StateType& src, StateType* dst) const override {
+    *dst += src;
+    return Status::OK();
+  }
+
+  Status Finalize(const StateType& src, Datum* output) const override {
+    *output = src.Finalize();
+    return Status::OK();
+  }
+
+  std::shared_ptr<DataType> out_type() const override { return StateType::out_type(); }
+
+ private:
+  StateType ConsumeDense(const ArrayType& array) const {
+    StateType local;
+
+    const auto values = array.raw_values();
+    const int64_t length = array.length();
+    for (int64_t i = 0; i < length; i++) {
+      local.sum += values[i];
+    }
+
+    local.count = length;
+
+    return local;
+  }
+
+  StateType ConsumeTiny(const ArrayType& array) const {
+    StateType local;
+
+    internal::BitmapReader reader(array.null_bitmap_data(), array.offset(),
+                                  array.length());
+    const auto values = array.raw_values();
+    for (int64_t i = 0; i < array.length(); i++) {
+      if (reader.IsSet()) {
+        local.sum += values[i];
+        local.count++;
+      }
+      reader.Next();
+    }
+
+    return local;
+  }
+
+  inline StateType UnrolledSum(uint8_t bits, const CType* values) const {
+    StateType local;
+
+    if (bits < 0xFF) {
+#define SUM_SHIFT(ITEM) values[ITEM] * static_cast<CType>(((bits >> ITEM) & 1U))
```

Comment:
I'm not seeing a reason that this can't be an inline function instead of a macro? Does it change the performance characteristics?

Reply:
I've added support for NaN handling at compile time via type traits and inline function. For some reason, using a type traits changes the behavior of clang optimizer and emits code with branches.

This does not affect gcc optimizer. For now, I'm going to leave this as-is since we're very likely to make an additional round on this code path to properly use vectorization.